src/main/scala/io/iohk/ethereum/network/TimeSlotStats.scala

```diff
@@ -0,0 +1,138 @@
+package io.iohk.ethereum.network
+
+import cats._
+import cats.implicits._
+import java.time.Clock
+import scala.concurrent.duration.{Duration, FiniteDuration}
+
+/** Track statistics over time a fixed size timewindow. */
+class TimeSlotStats[K, V: Monoid] private (
+    // Time resolution.
+    val slotDuration: FiniteDuration,
+    // The last written position in the buffer.
+    val lastIdx: Int,
+    // Ring buffer of slots statistics.
+    val buffer: TimeSlotStats.Buffer[K, V]
+)(implicit clock: Clock) {
+  import TimeSlotStats._
+
+  /** Overall length of the timewindow. */
+  def duration: FiniteDuration = slotDuration * slotCount
+  def slotCount: Int = buffer.size
+
+  /** Merge new stats for a given key in the current timestamp. */
+  def add(key: K, stat: V): TimeSlotStats[K, V] = {
+    val currSlotId = slotId(currentTimeMillis)
+    val lastEntry = buffer(lastIdx)
+
+    if (currSlotId == lastEntry.slotId) {
+      // We're still in the same timeslot, so just append stats.
+      val newEntry = lastEntry.add(key, stat)
+      updated(lastIdx, newEntry)
+    } else if (currSlotId > lastEntry.slotId) {
+      // Go to the next slot.
+      val newIdx = succ(lastIdx)
+      val newEntry = Entry(currSlotId, Map(key -> stat))
+      updated(newIdx, newEntry)
+    } else {
+      // Going backwards in time, just ignore it.
+      this
+    }
+  }
+
+  /** Forget all statistics about a given key. */
+  def remove(key: K): TimeSlotStats[K, V] =
+    updated(lastIdx, buffer.mapValues(_.remove(key)))
+
+  /** Aggregate stats for a key in all slots that are within the duration. */
+  def get(key: K, window: Option[Duration] = None): V =
+    fold(Monoid[V].empty, window getOrElse duration) { case (acc, stats) =>
+      stats.get(key).map(acc |+| _).getOrElse(acc)
+    }
+
+  /** Aggregate all stats in all slots within the duration. */
+  def getAll(window: Option[Duration] = None): Map[K, V] =
+    fold(Map.empty[K, V], window getOrElse duration) { case (acc, stats) =>
+      acc |+| stats
+    }
+
+  private def fold[A](init: A, window: Duration)(f: (A, Map[K, V]) => A) = {
+    val (start, end) = slotRange(currentTimeMillis, window)
+
+    def loop(idx: Int, acc: List[Map[K, V]]): List[Map[K, V]] = {
```

Comment:
Can this method be `@tailrec`?

Reply:
It should be. I can add it in the other PR just in case.

Reply:
Done.